### DISCUSSION TITLE
Errors when running GameControllerTest

### DISCUSSION THREAD

**Discussion by wsongah** [CastleLab/COMP3021-2021Fall-PA1-Student-Version#64] [Q&A]
I got an error regarding line126:  "final var wallPos = new Position(1, 1).offsetBy(direction.getOffset());"  when running one of the GameControllerTest Test: 
@EnumSource(value = Direction.class)
"@DisplayName("Make Move - Move to Adjacent Wall, Unlimited Lives")
    void testMakeMoveToAdjacentWallUnlimitedLives(final Direction direction)"
The error message says that: 
Position coordinates cannot be of a negative value.
java.lang.IllegalArgumentException: Position coordinates cannot be of a negative value.

But it seems to me that negative valuables are unavoidable for direction.getOffset(), for you have to test all four directions.
And "Position coordinates cannot be of a negative value." are also a restriction that cannot be deleted, right?
So, what's wrong with my understanding? How to fix this?
Thank you!

**Comment by Derppening** [comment]
Why would there be negative values? The single-unit offset of (1,1) is one of (0,1), (1,0), (2,1) or (1,2).

None of the offsets should result in a negative coordinate value.

**Reply by wsongah** [reply]
In the direction.java I got this:
 @NotNull
    public PositionOffset getOffset() {
        return new PositionOffset(getRowOffset(), getColOffset());
    }

    /**
     * @return The single-cell row offset of this direction.
     */
    public int getRowOffset() {
        return switch (this) {
            case UP -> -1;
            case DOWN -> 1;
            case LEFT, RIGHT -> 0;
        };
    }

    /**
     * @return The single-cell column offset of this direction.
     */
    public int getColOffset() {
        return switch (this) {
            case LEFT -> -1;
            case RIGHT -> 1;
            case UP, DOWN -> 0;
        };
    }
}
Shouldn't the 4 possible PositionOffset be (1,0), (-1,0), (0,1), (0,-1)?
Why should they be (0,1), (1,0), (2,1) or (1,2)?
I don't think I changed anything in Direction.java.

**Reply by Derppening** [reply]
The offsets themselves (i.e. `PositionOffset`) are (1,0), (-1,0), (0,1), (0,-1). There is no constraint that `PositionOffset` values must be non-negative, only `Position` values must be non-negative.

**Reply by Derppening** [reply]
> Why should they be (0,1), (1,0), (2,1) or (1,2)?

Because the original position is at (1,1), and by applying one of the above offsets, you will get (0,1) or (1,0) or (2,1) or (1,2).

**Reply by wsongah** [reply]
Thank you!
